pkg/util/http.go

```diff
@@ -195,76 +198,75 @@ func (e MsgSizeTooLargeErr) Is(err error) bool {
-			err = MsgSizeTooLargeErr{Actual: len(body), Limit: maxSize}
-		}
-	}()
+func decompressRequest(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
 	if expectedSize > maxSize {
 		return nil, MsgSizeTooLargeErr{Actual: expectedSize, Limit: maxSize}
 	}
-	buffer, ok := tryBufferFromReader(reader)
-	if ok {
-		body, err = decompressFromBuffer(dst, buffer, maxSize, compression, sp)
-		return
+	if compression != NoCompression && compression != RawSnappy && compression != Gzip {
+		return nil, fmt.Errorf("unrecognized compression type %v", compression)
+	}
+
+	if compression == NoCompression || compression == RawSnappy {
+		buf, ok := tryBufferFromReader(reader)
+		if ok {
+			return decompressFromBuffer(dst, buf, maxSize, compression, sp)
+		}
+	}
+
+	if sp != nil {
+		sp.LogFields(otlog.Event("util.ParseProtoReader[decompress]"), otlog.Int("expectedSize", expectedSize))
+	}
+
+	if compression == Gzip {
+		var err error
+		reader, err = gzip.NewReader(reader)
+		if err != nil {
+			return nil, errors.Wrap(err, "create gzip reader")
+		}
 	}
-	body, err = decompressFromReader(dst, reader, expectedSize, maxSize, compression, sp)
-	return
-}
 
-func decompressFromReader(dst []byte, reader io.Reader, expectedSize, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
-	var (
-		buf  bytes.Buffer
-		body []byte
-		err  error
-	)
+	// Limit at maxSize+1 so we can tell when the size is exceeded
+	reader = io.LimitReader(reader, int64(maxSize)+1)
+	var buf bytes.Buffer
 	if expectedSize > 0 {
 		buf.Grow(expectedSize + bytes.MinRead) // extra space guarantees no reallocation
 	}
-	// Read from LimitReader with limit max+1. So if the underlying
-	// reader is over limit, the result will be bigger than max.
-	reader = io.LimitReader(reader, int64(maxSize)+1)
-	switch compression {
-	case NoCompression:
-		_, err = buf.ReadFrom(reader)
-		body = buf.Bytes()
-	case RawSnappy:
-		_, err = buf.ReadFrom(reader)
-		if err != nil {
-			return nil, err
+	if _, err := buf.ReadFrom(reader); err != nil {
+		if compression == Gzip {
+			return nil, errors.Wrap(err, "decompress gzip")
 		}
-		body, err = decompressFromBuffer(dst, &buf, maxSize, RawSnappy, sp)
+		return nil, errors.Wrap(err, "read body")
 	}
-	return body, err
+
+	return decompressFromBuffer(dst, &buf, maxSize, compression, sp)
 }
 
 func decompressFromBuffer(dst []byte, buffer *bytes.Buffer, maxSize int, compression CompressionType, sp opentracing.Span) ([]byte, error) {
-	if len(buffer.Bytes()) > maxSize {
-		return nil, MsgSizeTooLargeErr{Actual: len(buffer.Bytes()), Limit: maxSize}
+	if buffer.Len() > maxSize {
+		return nil, MsgSizeTooLargeErr{Actual: -1, Limit: maxSize}
 	}
-	switch compression {
-	case NoCompression:
+
+	if sp != nil {
+		sp.LogFields(otlog.Event("util.ParseProtoReader[decompress]"), otlog.Int("size", buffer.Len()))
+	}
+
+	if compression != RawSnappy {
```

Comment:
nit: I'd be explicit about which compressions can use this approach, ie. None and Gzip.

```suggestion
	if compression == NoCompression || compression == Gzip {
```